File: src/memento/graph_debug/layout.py
from __future__ import annotations

import hashlib
import math
from collections import Counter, defaultdict
from collections.abc import Iterable
from dataclasses import dataclass

from memento.graph_debug.models import GraphEdge, GraphNode, GraphPosition

_LAYOUT_VERSION = "v1"


@dataclass(frozen=True, slots=True)
class AggregateNode:
    id: str
    label: str
    namespace: str
    member_count: int
    markdown_bytes: int
    asset_bytes: int
    combined_bytes: int
    explicit_in_degree: int
    explicit_out_degree: int
    broken_link_count: int
    orphan_count: int
    type_counts: tuple[tuple[str, int], ...]
    status_counts: tuple[tuple[str, int], ...]
    coarse_position: GraphPosition


@dataclass(frozen=True, slots=True)
class AggregateEdge:
    id: str
    source: str
    target: str
    explicit_edge_count: int
    canonical: bool = True


@dataclass(frozen=True, slots=True)
class AggregateLayout:
    seed: str
    version: str
    clusters: tuple[AggregateNode, ...]
    edges: tuple[AggregateEdge, ...]
    memberships: tuple[tuple[str, str], ...]


def aggregate_layout(
    nodes: Iterable[GraphNode],
    edges: Iterable[GraphEdge],
    *,
    repository_revision: str,
    cluster_limit: int,
) -> AggregateLayout:
    ordered_nodes = tuple(sorted(nodes, key=lambda item: item.id))
    ordered_edges = tuple(sorted(edges, key=lambda item: item.id))
    by_namespace: dict[str, list[GraphNode]] = defaultdict(list)
    for node in ordered_nodes:
        by_namespace[node.namespace].append(node)

    groups: list[tuple[str, tuple[GraphNode, ...]]] = []
    for namespace, members in sorted(by_namespace.items()):
        components = _explicit_components(tuple(members), ordered_edges)
        for index, component in enumerate(components):
            cluster_id = _cluster_id(namespace, index, component)
            groups.append((cluster_id, component))
    if len(groups) > cluster_limit:
        groups = _merge_small_groups(groups, cluster_limit)

    memberships = {node.id: cluster_id for cluster_id, members in groups for node in members}
    clusters = tuple(
        _aggregate_node(
            cluster_id,
            members,
            repository_revision=repository_revision,
            ordinal=index,
            total=len(groups),
        )
        for index, (cluster_id, members) in enumerate(groups)
    )
    counts: Counter[tuple[str, str]] = Counter()
    for edge in ordered_edges:
        if edge.target is None:
            continue
        source = memberships.get(edge.source)
        target = memberships.get(edge.target)
        if source is None or target is None or source == target:
            continue
        counts[(source, target)] += 1
    aggregate_edges = tuple(
        AggregateEdge(
            id=f"cluster-explicit:{source}:{target}",
            source=source,
            target=target,
            explicit_edge_count=count,
        )
        for (source, target), count in sorted(counts.items())
    )
    return AggregateLayout(
        seed=repository_revision,
        version=_LAYOUT_VERSION,
        clusters=clusters,
        edges=aggregate_edges,
        memberships=tuple(sorted(memberships.items())),
    )


def _explicit_components(
    members: tuple[GraphNode, ...], edges: tuple[GraphEdge, ...]
) -> tuple[tuple[GraphNode, ...], ...]:
    ids = {node.id for node in members}
    neighbours: dict[str, set[str]] = {node_id: set() for node_id in ids}
    for edge in edges:
        if edge.target is not None and edge.source in ids and edge.target in ids:
            neighbours[edge.source].add(edge.target)
            neighbours[edge.target].add(edge.source)
    node_by_id = {node.id: node for node in members}
    components = []
    unseen = set(ids)
    while unseen:
        start = min(unseen)
        stack = [start]
        component = []
        unseen.remove(start)
        while stack:
            current = stack.pop()
            component.append(node_by_id[current])
            for neighbour in sorted(neighbours[current], reverse=True):
                if neighbour in unseen:
                    unseen.remove(neighbour)
                    stack.append(neighbour)
        components.append(tuple(sorted(component, key=lambda item: item.id)))
    return tuple(sorted(components, key=lambda component: component[0].id))


def _merge_small_groups(
    groups: list[tuple[str, tuple[GraphNode, ...]]], cluster_limit: int
) -> list[tuple[str, tuple[GraphNode, ...]]]:
    retained = sorted(groups, key=lambda item: (-len(item[1]), item[0]))[: cluster_limit - 1]
    retained_ids = {item[0] for item in retained}
    overflow = tuple(
        sorted(
            (
                node
                for cluster_id, members in groups
                if cluster_id not in retained_ids
                for node in members
            ),
            key=lambda item: item.id,
        )
    )
    retained.append(("cluster:overflow", overflow))
    return sorted(retained, key=lambda item: item[0])


def _cluster_id(namespace: str, index: int, members: tuple[GraphNode, ...]) -> str:
    digest = hashlib.sha256(
        (namespace + "\0" + "\0".join(node.id for node in members)).encode("utf-8")
    ).hexdigest()[:12]
    return f"cluster:{namespace.strip('/') or 'root'}:{index}:{digest}"


def _aggregate_node(
    cluster_id: str,
    members: tuple[GraphNode, ...],
    *,
    repository_revision: str,
    ordinal: int,
    total: int,
) -> AggregateNode:
    namespace = members[0].namespace if members else "/"
    return AggregateNode(
        id=cluster_id,
        label=namespace if len(members) != 1 else members[0].title,
        namespace=namespace,
        member_count=len(members),
        markdown_bytes=sum(node.markdown_bytes for node in members),
        asset_bytes=sum(node.asset_bytes for node in members),
        combined_bytes=sum(node.combined_bytes for node in members),
        explicit_in_degree=sum(node.explicit_in_degree for node in members),
        explicit_out_degree=sum(node.explicit_out_degree for node in members),
        broken_link_count=sum(node.broken_link_count for node in members),
        orphan_count=sum(node.orphan for node in members),
        type_counts=tuple(sorted(Counter(node.type for node in members).items())),
        status_counts=tuple(sorted(Counter(node.status for node in members).items())),
        coarse_position=_cluster_position(
            cluster_id,
            repository_revision=repository_revision,
            ordinal=ordinal,
            total=total,
        ),
    )


def _cluster_position(
    cluster_id: str, *, repository_revision: str, ordinal: int, total: int
) -> GraphPosition:
    digest = hashlib.sha256(
        f"{_LAYOUT_VERSION}\0{repository_revision}\0{cluster_id}".encode()
    ).digest()
    jitter = int.from_bytes(digest[:4], "big") / 2**32 - 0.5
    angle = math.tau * (ordinal + 0.5 + jitter * 0.2) / max(1, total)
    radius = 3.0 + math.sqrt(max(1, total)) * 0.6
    z = (int.from_bytes(digest[4:8], "big") / 2**32 - 0.5) * 2.0
    return GraphPosition(x=math.cos(angle) * radius, y=math.sin(angle) * radius, z=z)

File: internal/graphdebug/layout.go
package graphdebug

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"math"
	"sort"
	"strings"
)

type CountPair [2]any
type AggregateNode struct {
	ID                string      `json:"id"`
	Label             string      `json:"label"`
	Namespace         string      `json:"namespace"`
	MemberCount       int         `json:"member_count"`
	MarkdownBytes     int64       `json:"markdown_bytes"`
	AssetBytes        int64       `json:"asset_bytes"`
	CombinedBytes     int64       `json:"combined_bytes"`
	ExplicitInDegree  int         `json:"explicit_in_degree"`
	ExplicitOutDegree int         `json:"explicit_out_degree"`
	BrokenLinkCount   int         `json:"broken_link_count"`
	OrphanCount       int         `json:"orphan_count"`
	TypeCounts        []CountPair `json:"type_counts"`
	StatusCounts      []CountPair `json:"status_counts"`
	CoarsePosition    Position    `json:"coarse_position"`
}
type AggregateEdge struct {
	ID                string   `json:"id"`
	Source            string   `json:"source"`
	Target            string   `json:"target"`
	ExplicitEdgeCount int      `json:"explicit_edge_count"`
	Kind              string   `json:"kind"`
	Canonical         bool     `json:"canonical"`
	Similarity        *float64 `json:"similarity"`
}
type Layout struct {
	Seed        string          `json:"seed"`
	Version     string          `json:"version"`
	Clusters    []AggregateNode `json:"clusters"`
	Edges       []AggregateEdge `json:"edges"`
	Memberships []CountPair     `json:"memberships"`
}
type nodeGroup struct {
	id    string
	nodes []Node
}

func explicitComponents(nodes []Node, edges []Edge) [][]Node {
	ids := map[string]bool{}
	byID := map[string]Node{}
	neighbors := map[string]map[string]bool{}
	for _, n := range nodes {
		ids[n.ID] = true
		byID[n.ID] = n
		neighbors[n.ID] = map[string]bool{}
	}
	for _, e := range edges {
		if e.Kind == "explicit" && e.Target != nil && ids[e.Source] && ids[*e.Target] {
			neighbors[e.Source][*e.Target] = true
			neighbors[*e.Target][e.Source] = true
		}
	}
	unseen := map[string]bool{}
	for id := range ids {
		unseen[id] = true
	}
	out := [][]Node{}
	for len(unseen) > 0 {
		start := ""
		for id := range unseen {
			if start == "" || id < start {
				start = id
			}
		}
		delete(unseen, start)
		stack := []string{start}
		component := []Node{}
		for len(stack) > 0 {
			current := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			component = append(component, byID[current])
			next := []string{}
			for id := range neighbors[current] {
				if unseen[id] {
					next = append(next, id)
				}
			}
			sort.Sort(sort.Reverse(sort.StringSlice(next)))
			for _, id := range next {
				delete(unseen, id)
				stack = append(stack, id)
			}
		}
		sort.Slice(component, func(i, j int) bool { return component[i].ID < component[j].ID })
		out = append(out, component)
	}
	sort.Slice(out, func(i, j int) bool { return out[i][0].ID < out[j][0].ID })
	return out
}
func sparseNamespace(nodes []Node, components [][]Node) bool {
	if len(nodes) <= 1 || len(components) <= 1 {
		return false
	}
	single := 0
	for _, c := range components {
		if len(c) == 1 {
			single++
		}
	}
	return float64(single)/float64(len(components)) >= .5
}
func clusterID(namespace string, index int, nodes []Node) string {
	input := namespace + "\x00"
	for i, n := range nodes {
		if i > 0 {
			input += "\x00"
		}
		input += n.ID
	}
	sum := sha256.Sum256([]byte(input))
	return "cluster:" + strings.Trim(namespace, "/") + ":" + integerText(int64(index)) + ":" + hex.EncodeToString(sum[:6])
}
func mergeGroups(groups []nodeGroup, limit int) []nodeGroup {
	sort.Slice(groups, func(i, j int) bool {
		if len(groups[i].nodes) != len(groups[j].nodes) {
			return len(groups[i].nodes) > len(groups[j].nodes)
		}
		return groups[i].id < groups[j].id
	})
	retained := append([]nodeGroup{}, groups[:max(0, limit-1)]...)
	keep := map[string]bool{}
	for _, g := range retained {
		keep[g.id] = true
	}
	overflow := []Node{}
	for _, g := range groups {
		if !keep[g.id] {
			overflow = append(overflow, g.nodes...)
		}
	}
	sort.Slice(overflow, func(i, j int) bool { return overflow[i].ID < overflow[j].ID })
	retained = append(retained, nodeGroup{"cluster:overflow", overflow})
	sort.Slice(retained, func(i, j int) bool { return retained[i].id < retained[j].id })
	return retained
}
func countPairs(values map[string]int) []CountPair {
	keys := []string{}
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	out := make([]CountPair, len(keys))
	for i, key := range keys {
		out[i] = CountPair{key, values[key]}
	}
	return out
}
func clusterPosition(id, revision string, ordinal, total int) Position {
	sum := sha256.Sum256([]byte("v1\x00" + revision + "\x00" + id))
	jitter := float64(binary.BigEndian.Uint32(sum[:4]))/math.Exp2(32) - .5
	angle := 2 * math.Pi * (float64(ordinal) + .5 + jitter*.2) / float64(max(1, total))
	radius := 3 + math.Sqrt(float64(max(1, total)))*.6
	z := (float64(binary.BigEndian.Uint32(sum[4:8]))/math.Exp2(32) - .5) * 2
	return Position{math.Cos(angle) * radius, math.Sin(angle) * radius, z}
}
func aggregateNode(group nodeGroup, revision string, ordinal, total int) AggregateNode {
	namespace := "/"
	if len(group.nodes) > 0 {
		namespace = group.nodes[0].Namespace
	}
	label := namespace
	if namespace == "/trash/" {
		label = "Trash"
	} else if len(group.nodes) == 1 {
		label = group.nodes[0].Title
	}
	out := AggregateNode{ID: group.id, Label: label, Namespace: namespace, MemberCount: len(group.nodes), CoarsePosition: clusterPosition(group.id, revision, ordinal, total)}
	types, statuses := map[string]int{}, map[string]int{}
	for _, n := range group.nodes {
		out.MarkdownBytes += n.MarkdownBytes
		out.AssetBytes += n.AssetBytes
		out.CombinedBytes += n.CombinedBytes
		out.ExplicitInDegree += n.ExplicitInDegree
		out.ExplicitOutDegree += n.ExplicitOutDegree
		out.BrokenLinkCount += n.BrokenLinkCount
		if n.Orphan {
			out.OrphanCount++
		}
		types[n.Type]++
		statuses[n.Status]++
	}
	out.TypeCounts = countPairs(types)
	out.StatusCounts = countPairs(statuses)
	return out
}
func AggregateLayout(nodes []Node, edges []Edge, revision string, clusterLimit int) Layout {
	nodes = append([]Node{}, nodes...)
	edges = append([]Edge{}, edges...)
	sort.Slice(nodes, func(i, j int) bool { return nodes[i].ID < nodes[j].ID })
	sort.Slice(edges, func(i, j int) bool { return edges[i].ID < edges[j].ID })
	byNamespace := map[string][]Node{}
	for _, n := range nodes {
		byNamespace[n.Namespace] = append(byNamespace[n.Namespace], n)
	}
	namespaces := []string{}
	for n := range byNamespace {
		namespaces = append(namespaces, n)
	}
	sort.Strings(namespaces)
	groups := []nodeGroup{}
	for _, namespace := range namespaces {
		members := byNamespace[namespace]
		components := explicitComponents(members, edges)
		if namespace == "/trash/" || sparseNamespace(members, components) {
			id := clusterID(namespace, 0, members)
			if namespace == "/trash/" {
				id = "cluster:trash"
			}
			groups = append(groups, nodeGroup{id, members})
		} else {
			for i, component := range components {
				groups = append(groups, nodeGroup{clusterID(namespace, i, component), component})
			}
		}
	}
	if len(groups) > clusterLimit {
		trash := []nodeGroup{}
		others := []nodeGroup{}
		for _, g := range groups {
			if g.id == "cluster:trash" {
				trash = append(trash, g)
			} else {
				others = append(others, g)
			}
		}
		if len(trash) > 0 && clusterLimit > 1 {
			groups = append(trash, mergeGroups(others, clusterLimit-1)...)
		} else {
			groups = mergeGroups(groups, clusterLimit)
		}
	}
	memberships := map[string]string{}
	clusters := make([]AggregateNode, len(groups))
	for i, g := range groups {
		clusters[i] = aggregateNode(g, revision, i, len(groups))
		for _, n := range g.nodes {
			memberships[n.ID] = g.id
		}
	}
	type key struct{ kind, source, target string }
	counts := map[key]int{}
	sums := map[key]float64{}
	for _, e := range edges {
		if e.Target == nil {
			continue
		}
		source, target := memberships[e.Source], memberships[*e.Target]
		if source == "" || target == "" || source == target {
			continue
		}
		k := key{e.Kind, source, target}
		counts[k]++
		if e.Kind == "semantic_similarity" && e.Similarity != nil {
			sums[k] += *e.Similarity
		}
	}
	keys := []key{}
	for k := range counts {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool {
		if keys[i].kind != keys[j].kind {
			return keys[i].kind < keys[j].kind
		}
		if keys[i].source != keys[j].source {
			return keys[i].source < keys[j].source
		}
		return keys[i].target < keys[j].target
	})
	aggregateEdges := []AggregateEdge{}
	for _, k := range keys {
		count := counts[k]
		edge := AggregateEdge{ID: "cluster-" + k.kind + ":" + k.source + ":" + k.target, Source: k.source, Target: k.target, Kind: k.kind, Canonical: k.kind == "explicit"}
		if k.kind == "explicit" {
			edge.ExplicitEdgeCount = count
		}
		if k.kind == "semantic_similarity" {
			value := sums[k] / float64(count)
			edge.Similarity = &value
		}
		aggregateEdges = append(aggregateEdges, edge)
	}
	pairs := []CountPair{}
	ids := []string{}
	for id := range memberships {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	for _, id := range ids {
		pairs = append(pairs, CountPair{id, memberships[id]})
	}
	return Layout{revision, "v1", clusters, aggregateEdges, pairs}
}
